nearcore/src/config.rs

```diff
@@ -1327,39 +1336,93 @@ pub fn load_config(
-        format!("Failed reading node key file from {}", node_key_path.display())
-    })?;
+    let network_signer_result = NodeKeyFile::from_file(&node_key_path);
+    let network_signer = match network_signer_result {
+        Ok(node_key_file) => Some(node_key_file),
+        Err(_) => {
+            validation_errors.push_errors(ValidationError::NodeKeyFileError {
+                error_message: format!(
+                    "Failed reading node key file from {}",
+                    node_key_path.display()
+                ),
+            });
+            None
+        }
+    };
+
+    let genesis_file = dir.join(&config.genesis_file);
+    let genesis_result = match &config.genesis_records_file {
+        // only load Genesis from file. Skip test for now.
+        // this allows us to know the chain_id in order to check tracked_shards even if semantics checks fail.
+        Some(records_file) => Genesis::from_files(
+            &genesis_file,
+            &dir.join(records_file),
+            GenesisValidationMode::UnsafeFast,
+        ),
+        None => Genesis::from_file(&genesis_file, GenesisValidationMode::UnsafeFast),
+    };
 
-    let genesis = match &config.genesis_records_file {
-        Some(records_file) => {
-            Genesis::from_files(&genesis_file, &dir.join(records_file), genesis_validation)
+    let genesis = match genesis_result {
+        Ok(genesis) => {
+            genesis
+                .validate(genesis_validation)
+                .map_or_else(|e| validation_errors.push_errors(e), |_| ());
+            if matches!(genesis.config.chain_id.as_ref(), "mainnet" | "testnet" | "betanet")
+                && config.tracked_shards.is_empty()
+            {
+                // Make sure validators tracks all shards, see
+                // https://github.com/near/nearcore/issues/7388
+                let error_message = format!("The `chain_id` field specified in genesis is among mainnet/betanet/testnet, so validator must track all shards. Please change `tracked_shards` field in config.json to be any non-empty vector");
+                validation_errors.push_errors(ValidationError::CrossFileSematicError {
+                    error_message: error_message,
+                });
+            }
+            Some(genesis)
+        }
+        Err(error) => {
+            validation_errors.push_errors(error);
+            None
         }
-        None => Genesis::from_file(&genesis_file, genesis_validation),
     };
 
-    if validator_signer.is_some()
-        && matches!(genesis.config.chain_id.as_ref(), "mainnet" | "testnet" | "betanet")
-    {
-        // Make sure validators tracks all shards, see
-        // https://github.com/near/nearcore/issues/7388
-        anyhow::ensure!(!config.tracked_shards.is_empty(),
-                        "Validator must track all shards. Please change `tracked_shards` field in config.json to be any non-empty vector");
-    }
+    validation_errors.panic_if_errors();
```

Comment:
why not just return an `Err(anyhow::Error)`? Could replace the `panic_if_errors()` function with a function that returns an error, maybe like:

```rust
    pub fn ok(&self) -> anyhow::Result<()> {
        match self.generate_error_message_per_type() {
            Some(e) => Err(anyhow::Error::msg(e)),
            None => Ok(()),
        }
    }
```

or called whatever makes more sense to you. I bring it up just because I think the codebase in general has a problem with panicking too much as an error handling method in situations where nothing unexpected has happened: https://github.com/near/nearcore/issues/5485